ark-core/src/vtxo.rs

```diff
@@ -320,3 +358,294 @@ where
+        &self.spend_info
+    }
+
+    /// Get the tweaked public key
+    pub fn tweaked_public_key(&self) -> bitcoin::key::TweakedPublicKey {
+        self.tweaked_public_key
+    }
+
+    /// Get the X-only public key from the tweaked public key
+    pub fn x_only_public_key(&self) -> XOnlyPublicKey {
+        self.tweaked_public_key.to_x_only_public_key()
+    }
+
+    /// Get the script pubkey (P2TR output script)
+    pub fn script_pubkey(&self) -> ScriptBuf {
+        tr_script_pubkey(&self.spend_info)
+    }
+
+    /// Create an ArkAddress from this script
+    pub fn ark_address(&self, network: Network, server_pubkey: XOnlyPublicKey) -> ArkAddress {
+        ArkAddress::new(network, server_pubkey, self.tweaked_public_key)
+    }
+
+    /// Get the on-chain address for this script
+    pub fn onchain_address(&self, network: Network) -> Address {
+        Address::from_script(&self.script_pubkey(), network).expect("valid taproot script")
+    }
+
+    /// Find a leaf by script hex
+    pub fn find_leaf(&self, script_hex: &str) -> Result<(ScriptBuf, taproot::ControlBlock), Error> {
+        let target_script = ScriptBuf::from_bytes(
+            hex::decode(script_hex)
+                .map_err(|e| Error::ad_hoc(format!("failed to decode script hex: {e}")))?,
+        );
+
+        for script in &self.scripts {
+            if *script == target_script {
+                let control_block = self
+                    .spend_info
+                    .control_block(&(script.clone(), LeafVersion::TapScript))
+                    .ok_or_else(|| Error::ad_hoc("control block not found for script"))?;
+
+                return Ok((script.clone(), control_block));
+            }
+        }
+
+        Err(Error::ad_hoc(format!("leaf '{script_hex}' not found")))
+    }
+
+    /// Get control block for a specific script
+    pub fn control_block(&self, script: &ScriptBuf) -> Result<taproot::ControlBlock, Error> {
+        self.spend_info
+            .control_block(&(script.clone(), LeafVersion::TapScript))
+            .ok_or_else(|| Error::ad_hoc("control block not found for script"))
+    }
+
+    /// Get all tap leaves with their control blocks
+    pub fn tap_leaves(&self) -> Vec<(ScriptBuf, taproot::ControlBlock)> {
+        self.scripts
+            .iter()
+            .filter_map(|script| {
+                self.spend_info
+                    .control_block(&(script.clone(), LeafVersion::TapScript))
+                    .map(|control_block| (script.clone(), control_block))
+            })
+            .collect()
+    }
+}
+
+/// Encoded representation of a VirtualUtxoScript
+#[derive(Debug, Clone)]
+pub struct EncodedVirtualUtxoScript {
+    pub scripts: Vec<String>,
+}
+
+impl From<&VirtualUtxoScript> for EncodedVirtualUtxoScript {
+    fn from(vtxo_script: &VirtualUtxoScript) -> Self {
+        Self {
+            scripts: vtxo_script.encode(),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I haven't checked thoroughly, but I believe that some of these tests are testing APIs that are never used anywhere else. We should trim down the tests to whatever is relevant.